src/redismodule.h

```diff
@@ -257,6 +259,113 @@ typedef enum {
+    const char *changes;
+} RedisModuleCommandHistoryEntryV1;
+
+#define RedisModuleCommandHistoryEntry RedisModuleCommandHistoryEntryV1
+
+typedef struct RedisModuleCommandKeySpec {
+    uint64_t flags; /* REDISMODULE_CMD_KEY_* macros. */
+    RedisModuleKeySpecBeginSearchType begin_search_type;
+    union {
+        struct {
+            /* The index from which we start the search for keys */
+            int pos;
+        } index;
+        struct {
+            /* The keyword that indicates the beginning of key args */
+            const char *keyword;
+            /* An index in argv from which to start searching.
+             * Can be negative, which means start search from the end, in reverse
+             * (Example: -2 means to start in reverse from the panultimate arg) */
+            int startfrom;
+        } keyword;
+    } bs;
+    RedisModuleKeySpecFindKeysType find_keys_type;
+    union {
+        struct {
+            /* Index of the last key relative to the result of the begin search
+             * step. Can be negative, in which case it's not relative. -1
+             * indicating till the last argument, -2 one before the last and so
+             * on. */
+            int lastkey;
+            /* How many args should we skip after finding a key, in order to
+             * find the next one. */
+            int keystep;
+            /* If lastkey is -1, we use limit to stop the search by a factor. 0
+             * and 1 mean no limit. 2 means 1/2 of the remaining args, 3 means
+             * 1/3, and so on. */
+            int limit;
+        } range;
+        struct {
+            /* Index of the argument containing the number of keys to come
+             * relative to the result of the begin search step */
+            int keynumidx;
+            /* Index of the fist key. (Usually it's just after keynumidx, in
+             * which case it should be set to keynumidx + 1.) */
+            int firstkey;
+            /* How many args should we skip after finding a key, in order to
+             * find the next one, relative to the result of the begin search
+             * step. */
+            int keystep;
+        } keynum;
+    } fk;
+} RedisModuleCommandKeySpecV1;
+
+#define RedisModuleCommandKeySpec RedisModuleCommandKeySpecV1
+
+typedef struct RedisModuleCommandInfo {
+    const char *summary;          /* Summary of the command */
+    const char *complexity;       /* Complexity description */
+    const char *since;            /* Debut module version of the command */
+    RedisModuleCommandHistoryEntry *history; /* History */
+    /* An array of strings that are meant o be hints for clients/proxies
+     * regarding this command */
+    const char **hints;
+    /* Number of arguments, it is possible to use -N to say >= N */
+    int arity;
+    RedisModuleCommandKeySpec *key_specs;
+    RedisModuleCommandArg *args;
+} RedisModuleCommandInfoV1;
+
+#define RedisModuleCommandInfo RedisModuleCommandInfoV1
```

Comment:
I see no benefit whatsoever with this approach. The macros and types all disappear at compile time anyway. I suggest removing the V1 indirection. @guybe7 Did I miss anything?

Reply:
how can we maintain backward compatibility if we don't version the struct in redismodule.h

Reply:
there should be a `version` field at the beginning of each struct

Reply:
maybe we really don't need that #define (it's useful only if the writer is agnostic to the version and just passes the struct to the core. see `test_clientinfo`: if we didn't have that #define, the module would have to update the code in case it compiled with a newer redismodule.h, which has V2)

Reply:
in our case we fill the structs, so we have to be aware of the version we're using (if we want to work with a newer redismodule.h we would have to update the module code + make sure we work with a new-enough server)

Reply:
but anyway, we must provide the version of each struct we pass to redis

Reply:
When we add V2, we can rename the structs to xxxV1 and add that `#define RedisModuleCommandInfo RedisModuleCommandInfoV1`, can't we?

Reply:
If the same API should be able to accept both InfoV1 and InfoV2 structs, we need something inside the struct to tell the version, e.g. something like this: (don't we?)

```C
struct Info {
    int version;
    union {
        InfoV1 v1;
        InfoV2 v2;
    } versioned_info;
}
```

Reply:
either that or the API will take `void*` and assume the first 8 bytes are the `version`

Reply:
... like the Event structures which have a `1` inside them:

```C
static const RedisModuleEvent
    RedisModuleEvent_ReplicationRoleChanged = {
        REDISMODULE_EVENT_REPLICATION_ROLE_CHANGED,
        1
    },
```

Reply:
If we add a version tag inside the struct, I think we only need it for the top-level struct RedisModuleInfo. Everything inside it can be assumed to be V1, don't you think?

Another thing: Do we still want to support V1 in the future, so V2 is opt-in? If yes, then we can let RedisModuleInfo always be V1 and use the new name RedisModuleInfoV2 for the new one...?

Reply:
The version tag makes it a bit awkward to write info literals, especially if we need a version tag in every struct in a nested struct literal.

An alternative is to add another API when we need to add more info (e.g. `RM_SetCommandMoreInfo`). It's slightly ugly but it's very clear in terms of backward compatibility and of what's availability in each Redis version.

Reply:
I do agree that any struct we take from the user should have a version field as a first field, and that in this case it should only be present in the top level struct we take as pointer, and probably not in any of the nested structs.

as for the define, the only reason to use a define is if we want modules to by default use a struct name without version, so then when they compile with a newer version of the header, they'll implicitly use the new struct and won't need to manually edit their code to use the new one (change "v1" suffix to "v2" suffix).
in that case, they're also expected to populate the version field with some "dynamic" define (e.g. `REDISMODULE_TYPE_METHOD_VERSION`).

the downside of this is that as soon as the module compiles with the new header, it is automatically bound to the new struct and looses backwards compatibility with old redis versions.
if we don't go with that `define` approach, then a module will only move to `v2` when there's a field it wants to use, that didn't exist in v1.

i think in this case we should model this after other input structs we get (i.e. `RedisModuleTypeMethods`, not `RedisModuleClientInfo`)

WDYT?

Reply:
> I do agree that any struct we take from the user should have a version field as a first field, and that in this case it should only be present in the top level struct we take as pointer, and probably not in any of the nested structs.

@oranagra Fair enough. It's a good compromise.

Observations about RedisModuleTypeMethods:

* RedisModuleTypeMethods has a version field.
* RedisModuleTypeMethods has no version in it's name (defined nor typedeffed).
* The struct pointer parameter has type `void *` (in `RM_CreateDataType`).
* The version macro is changed "synchronistically" when Redis is updated.

    <details>
    <summary>"syncronistically"</summary>

    ```C
    /* Version of the RedisModuleTypeMethods structure. Once the RedisModuleTypeMethods 
     * structure is changed, this version number needs to be changed synchronistically. */
    #define REDISMODULE_TYPE_METHOD_VERSION 4
    ```
    </details>

* `RM_CreateDataType` is backward compatible with modules compiled with older versions of this struct, because the size of the struct doesn't matter. The version field is checked before fields added in version X are accessed.
* `RM_CreateDataType` is forward compatible. Future (unknown) versions are accepted and their fields simply have no effect on older Redis. Version checks use `>=` and allow the version to be any number > 0.

The forward/backward compatibilty relies on the fact that there may be unused/unknown fields in the end of the struct. This doesn't work for arrays of structs such as RedisModuleCommandKeySpec (which are packed side-by-side in memory) but it works for the top-level RedisModuleCommandInfo, which is what matters.

I think we shall adapt the same approach this approach, apart from the type `void *` which I don't see the point of.

Reply:
ok. so the top level struct will be backwards compatible like the DataType struct, and any nested structs in it can probably be handled using the version field of the top level one.
if / when we made changes to any of these structs, i guess we can always decide to keep the old struct definitions if we want new modules, that use the new header, keep using the old API. or we can force them to use the new one.
either way, redis has a way to be ABI compatible with old ones.

i didn't invest enough time thinking about this.. hope i'm not wrong.
maybe we can make an experiment later (inventing some additional field and see what it leads to)

Reply:
Done: Version field added, defines for struct names removed.

Reply:
just to make sure we are on the same page:
1. in a year from now we decide that we need some `newmember` in `keyword` in  `RedisModuleCommandKeySpec`
2. we need to define another struct `RedisModuleCommandKeySpec2` with `newmember`
3. we need to add `RedisModuleCommandKeySpec *key_specs2` to `RedisModuleCommandInfo` (Q1: we decided that `RedisModuleCommandInfo` can only expand, we can't change the already-existing content right?)
4. now let's imagine two situations:
   - module that compiled with old redismodule.h runs with new server: the server takes the struct from user (where the last accessible member is `args`), sees the version is `1` and never tries to access `key_specs2`
   - module that compiled with new redismodule.h runs with old server: the module passes a struct with `key_specs2` and version `2` but the server is only aware of version `1` so it's not aware that there's anything after `args`
   
questions/thoughts that require your reply:
1. Q1 above ^
2. is the scenario above describes our design?
3. maybe we should document it in the code? to never modify `RedisModuleCommandInfo`, just add stuff if needed